### ISSUE TITLE
请教微调相关的问题

### ISSUE THREAD

**Issue by githubxun** [shiyu-coder/Kronos#104]
你好，我使用2021 ~ 2025年的历史K线数据进行微调，发现 train_tokenizer 不能正常的收敛，这可能是哪里的问题呢 ？
部分日志如下：

--- Epoch 25/30 Summary ---
Validation Loss: 0.0048
Time This Epoch: 0:01:02
Total Time Elapsed: 0:26:33

[Rank 0, Epoch 26/30, Step 100/500] LR 0.000013, Loss: -0.0290
[Rank 0, Epoch 26/30, Step 200/500] LR 0.000012, Loss: -0.0297
[Rank 0, Epoch 26/30, Step 300/500] LR 0.000011, Loss: -0.0292
[Rank 0, Epoch 26/30, Step 400/500] LR 0.000010, Loss: -0.0293
[Rank 0, Epoch 26/30, Step 500/500] LR 0.000009, Loss: -0.0293

--- Epoch 26/30 Summary ---
Validation Loss: 0.0048
Time This Epoch: 0:01:02
Total Time Elapsed: 0:27:36

[Rank 0, Epoch 27/30, Step 100/500] LR 0.000008, Loss: -0.0282
[Rank 0, Epoch 27/30, Step 200/500] LR 0.000007, Loss: -0.0291
[Rank 0, Epoch 27/30, Step 300/500] LR 0.000007, Loss: -0.0286
[Rank 0, Epoch 27/30, Step 400/500] LR 0.000006, Loss: -0.0289
[Rank 0, Epoch 27/30, Step 500/500] LR 0.000005, Loss: -0.0284

--- Epoch 27/30 Summary ---
Validation Loss: 0.0050
Time This Epoch: 0:01:02
Total Time Elapsed: 0:28:38

[Rank 0, Epoch 28/30, Step 100/500] LR 0.000005, Loss: -0.0289
[Rank 0, Epoch 28/30, Step 200/500] LR 0.000004, Loss: -0.0285
[Rank 0, Epoch 28/30, Step 300/500] LR 0.000003, Loss: -0.0293
[Rank 0, Epoch 28/30, Step 400/500] LR 0.000003, Loss: -0.0296
[Rank 0, Epoch 28/30, Step 500/500] LR 0.000002, Loss: -0.0294

--- Epoch 28/30 Summary ---
Validation Loss: 0.0049
Time This Epoch: 0:01:02
Total Time Elapsed: 0:29:41

[Rank 0, Epoch 29/30, Step 100/500] LR 0.000002, Loss: -0.0288
[Rank 0, Epoch 29/30, Step 200/500] LR 0.000001, Loss: -0.0293
[Rank 0, Epoch 29/30, Step 300/500] LR 0.000001, Loss: -0.0294
[Rank 0, Epoch 29/30, Step 400/500] LR 0.000001, Loss: -0.0282
[Rank 0, Epoch 29/30, Step 500/500] LR 0.000001, Loss: -0.0288

--- Epoch 29/30 Summary ---
Validation Loss: 0.0049
Time This Epoch: 0:01:02
Total Time Elapsed: 0:30:43

[Rank 0, Epoch 30/30, Step 100/500] LR 0.000000, Loss: -0.0293
[Rank 0, Epoch 30/30, Step 200/500] LR 0.000000, Loss: -0.0287
[Rank 0, Epoch 30/30, Step 300/500] LR 0.000000, Loss: -0.0296
[Rank 0, Epoch 30/30, Step 400/500] LR 0.000000, Loss: -0.0291
[Rank 0, Epoch 30/30, Step 500/500] LR 0.000000, Loss: -0.0285

--- Epoch 30/30 Summary ---
Validation Loss: 0.0049
Time This Epoch: 0:01:02
Total Time Elapsed: 0:31:46

Training finished. Summary file saved.

**Comment by githubxun** [comment]
更重要的问题，我的这个Loss居然是一个负值！

**Comment by githubxun** [comment]
继续补充部分代码和日志信息如下：
train_tokenizer.py 文件144行增加日志输出
# Forward pass
zs, bsq_loss, _, _ = model(batch_x)
z_pre, z = zs

# Loss calculation
recon_loss_pre = F.mse_loss(z_pre, batch_x)
recon_loss_all = F.mse_loss(z, batch_x)
recon_loss = recon_loss_pre + recon_loss_all
print(recon_loss)
print(bsq_loss)

记录的部分日志如下：
recon_loss：
tensor(0.0389, device='cuda:3', grad_fn=<AddBackward0>)
tensor(0.0393, device='cuda:1', grad_fn=<AddBackward0>)
tensor(0.0432, device='cuda:2', grad_fn=<AddBackward0>)
tensor(0.0375, device='cuda:0', grad_fn=<AddBackward0>)
tensor(-0.0698, device='cuda:3', grad_fn=<AddBackward0>)
tensor(-0.0700, device='cuda:1', grad_fn=<AddBackward0>)
tensor(-0.0699, device='cuda:2', grad_fn=<AddBackward0>)
tensor(-0.0698, device='cuda:0', grad_fn=<AddBackward0>)

bsq_loss：
tensor(0.0457, device='cuda:2', grad_fn=<AddBackward0>)
tensor(0.0399, device='cuda:0', grad_fn=<AddBackward0>)
tensor(-0.0699, device='cuda:1', grad_fn=<AddBackward0>)
tensor(-0.0699, device='cuda:3', grad_fn=<AddBackward0>)
tensor(-0.0699, device='cuda:2', grad_fn=<AddBackward0>)
tensor(-0.0698, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.0423, device='cuda:3', grad_fn=<AddBackward0>)
tensor(0.0400, device='cuda:1', grad_fn=<AddBackward0>)
tensor(0.0455, device='cuda:2', grad_fn=<AddBackward0>)
tensor(0.0398, device='cuda:0', grad_fn=<AddBackward0>)
tensor(-0.0699, device='cuda:3', grad_fn=<AddBackward0>)
tensor(-0.0699, device='cuda:1', grad_fn=<AddBackward0>)
tensor(-0.0699, device='cuda:2', grad_fn=<AddBackward0>)
tensor(-0.0698, device='cuda:0', grad_fn=<AddBackward0>)

**Comment by shiyu-coder** [comment]
你好，tokenizer的loss为负是正常的，不影响训练，具体的原因你可以参考BSQ的原论文。不能正常收敛是指？微调的话一般学习率比较小，训几个epoch就会收敛。我看你这里训到了25个epoch，这已经很多了，可能早已过拟合了。

**Comment by thatname** [comment]
predictor 无论怎么训练 loss 都大于3，这个很不正常吧？要怎么调整呢
